src/java/org/apache/cassandra/db/partitions/BTreePartitionUpdater.java

```diff
@@ -0,0 +1,174 @@
+
+    public BTreePartitionUpdater(MemtableAllocator allocator, OpOrder.Group writeOp, UpdateTransaction indexer)
+    {
+        this.allocator = allocator;
+        this.writeOp = writeOp;
+        this.indexer = indexer;
+        this.heapSize = 0;
+        this.dataSize = 0;
+    }
+
+    private Row.Builder builder(Clustering<?> clustering)
+    {
+        boolean isStatic = clustering == Clustering.STATIC_CLUSTERING;
+        // We know we only insert/update one static per PartitionUpdate, so no point in saving the builder
+        if (isStatic)
+            return allocator.rowBuilder(writeOp);
+
+        if (regularBuilder == null)
+            regularBuilder = allocator.rowBuilder(writeOp);
+        return regularBuilder;
+    }
+
+    public Row apply(Row insert)
+    {
+        Row data = Rows.copy(insert, builder(insert.clustering())).build();
+        indexer.onInserted(insert);
+
+        this.dataSize += data.dataSize();
+        allocated(data.unsharedHeapSizeExcludingData());
+        return data;
+    }
+
+    public Row apply(Row existing, Row update)
+    {
+        Row.Builder builder = builder(existing.clustering());
+        colUpdateTimeDelta = Math.min(colUpdateTimeDelta, Rows.merge(existing, update, builder));
+
+        Row reconciled = builder.build();
+
+        indexer.onUpdated(existing, reconciled);
+
+        dataSize += reconciled.dataSize() - existing.dataSize();
+        allocated(reconciled.unsharedHeapSizeExcludingData() - existing.unsharedHeapSizeExcludingData());
+
+        return reconciled;
+    }
+
+    private DeletionInfo apply(DeletionInfo existing, DeletionInfo update)
+    {
+        if (update.isLive() || !update.mayModify(existing))
+            return existing;
+
+        if (!update.getPartitionDeletion().isLive())
+            indexer.onPartitionDeletion(update.getPartitionDeletion());
+
+        if (update.hasRanges())
+            update.rangeIterator(false).forEachRemaining(indexer::onRangeTombstone);
+
+        // Like for rows, we have to clone the update in case internal buffers (when it has range tombstones) reference
+        // memory we shouldn't hold into. But we don't ever store this off-heap currently so we just default to the
+        // HeapAllocator (rather than using 'allocator').
+        DeletionInfo newInfo = existing.mutableCopy().add(update.copy(HeapAllocator.instance));
+        allocated(newInfo.unsharedHeapSize() - existing.unsharedHeapSize());
+        return newInfo;
+    }
+
+    public BTreePartitionData mergePartitions(BTreePartitionData current, final PartitionUpdate update)
+    {
+        if (current == null)
+        {
+            current = BTreePartitionData.EMPTY;
+            this.allocated(BTreePartitionData.UNSHARED_HEAP_SIZE);
+        }
+
+        try
+        {
+            indexer.start();
+
+            return makeMergedPartition(current, update);
+        }
+        finally
+        {
+            indexer.commit();
+            finish();
```

Comment:
rename `finish()` to sth along `reportAllocatedMemory()` ?

Reply:
Done